### PR TITLE
Project Report: A Learned Federated Query Compiler

### PR DESCRIPTION
**Pull request by gsvic**

### CODE REVIEW

**Comment by sampsyo** [comment]
Looks like Zola doesn't like the most missing a name—it should be `YYYY-MM-DD-name.md`. Can you please add that?

**Comment by gsvic** [comment]
> Looks like Zola doesn't like the most missing a name—it should be `YYYY-MM-DD-name.md`. Can you please add that?

sure, done!

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
++++
+title = "A Learned Query Compiler"
+[extra]
+latex = true
+[[extra.authors]]
+name = "Victor Giannakouris"
+link = "https://www.linkedin.com/in/vgian/"
++++
+
+# What was the goal?
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md]: TBH I would probably recommend that you don't reuse my questions as seconds in your blog post. Imagine instead that you are a reader who stumbles across the 6120 blog, knows something about compilers but nothing about cross-query databases, and wants to know why this is important. What is the first thing you would tell that reader?

(I think your first paragraph here is already pretty good at that; it just doesn't need a heading.)

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+Then, the join between `A` and the result of `B ⋈ C` could be executed in the federation engine. The reason that we need
+to adjust this number, is that if we push-down a large join that produces a relatively large intermediate result,
+fetching this result from the external engine to the federation engine might result into big overheads that will make
+the query execution slow. Using machine learning and also by verifying it empirically, we conclude for this benchmark that
+a `table_limit` of three works the best for most of the queries. Sticking to that parameter we can achieve run most of the
+queries to the ~ 60% of time, compared to the vanilla implementation of Spark SQL. However, our optimizer is still on a 
+very immature stage. We believe that our ongoing extension in which the `table_limit`, as well as the selection
+of the actual tables that will be included to a join will improve even better improvements.
+
+## Conclusions and Feature Steps
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md]: Feature -> future?

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+[extra]
+latex = true
+[[extra.authors]]
+name = "Victor Giannakouris"
+link = "https://www.linkedin.com/in/vgian/"
++++
+
+# What was the goal?
+The goal of this project was to create an optimizer for a cross-database query compiler. A cross-database query, also 
+known as federated query, is an SQL query that consists of tables that span across different, individual database
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md]: Consider italicizing new terms the first time you introduce them, like *federated query*.

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+# What was the goal?
+The goal of this project was to create an optimizer for a cross-database query compiler. A cross-database query, also 
+known as federated query, is an SQL query that consists of tables that span across different, individual database 
+systems. The main weakness with the already existing federated query optimization approaches is of the complexity
+of integration, as the existing solutions need to obtain specific knowledge from the externally connected databases
+by writing complex drivers. We propose a solution based on a machine learning model that enables the easy integration of
+any SQL-based system, with less than 10 loc.
+
+## Federated Query Compilation Challenges
+A lot of interesting challenges are arising in cross-database query compilation, due to the heterogeneity of the underlying
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md]: are arising -> arise

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+The goal of this project was to create an optimizer for a cross-database query compiler. A cross-database query, also 
+known as federated query, is an SQL query that consists of tables that span across different, individual database 
+systems. The main weakness with the already existing federated query optimization approaches is of the complexity
+of integration, as the existing solutions need to obtain specific knowledge from the externally connected databases
+by writing complex drivers. We propose a solution based on a machine learning model that enables the easy integration of
+any SQL-based system, with less than 10 loc.
+
+## Federated Query Compilation Challenges
+A lot of interesting challenges are arising in cross-database query compilation, due to the heterogeneity of the underlying 
+infrastructure. An interesting challenge that arises is the following: How could the query plan be split to further subqueries and how the execution of these subqueries could be orchestrated across the different data sources in order to utilize the external data systems and take advantage of their powerful implementations (e.g. access methods and join algorithms)?
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md]: This last sentence gets a little long and hard to follow. Maybe it should be 2–3 separate questions?

Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md]: Add a comma after "e.g." https://capra.cs.cornell.edu/styleguide/#egie

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+of integration, as the existing solutions need to obtain specific knowledge from the externally connected databases
+by writing complex drivers. We propose a solution based on a machine learning model that enables the easy integration of
+any SQL-based system, with less than 10 loc.
+
+## Federated Query Compilation Challenges
+A lot of interesting challenges are arising in cross-database query compilation, due to the heterogeneity of the underlying 
+infrastructure. An interesting challenge that arises is the following: How could the query plan be split to further subqueries and how the execution of these subqueries could be orchestrated across the different data sources in order to utilize the external data systems and take advantage of their powerful implementations (e.g. access methods and join algorithms)? 
+
+In this project, we experimented with a cross-database query system called Spark SQL, and we developed a cross-database query compiler that takes into account the 
+heterogeneity of the underlying infrastructure and it's able to optimize the query with respect to each table location.
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md]: it's able -> its ability? Not 100% sure what you meant here.

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+# What did you do?
+We mainly focused on the optimizer of our federated query compiler. We implemented a federated query compiler which is
+able to do optimize federated queries by splitting the execution of the initial query into individual, smaller parts
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:31]: "to do optimize"

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+name = "Victor Giannakouris"
+link = "https://www.linkedin.com/in/vgian/"
++++
+
+# What was the goal?
+The goal of this project was to create an optimizer for a cross-database query compiler. A cross-database query, also 
+known as federated query, is an SQL query that consists of tables that span across different, individual database 
+systems. The main weakness with the already existing federated query optimization approaches is of the complexity
+of integration, as the existing solutions need to obtain specific knowledge from the externally connected databases
+by writing complex drivers. We propose a solution based on a machine learning model that enables the easy integration of
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md]: I think this sentence could be even clearer about what the problem is. Namely, this kind of design fundamentally needs to know things about the underlying engines: table sizes, relative storage speeds, etc. Existing systems get those manually. We want to get them automatically. (But giving a clear idea of _what_ that integration must consist of would help a lot.)

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+A lot of interesting challenges are arising in cross-database query compilation, due to the heterogeneity of the underlying 
+infrastructure. An interesting challenge that arises is the following: How could the query plan be split to further subqueries and how the execution of these subqueries could be orchestrated across the different data sources in order to utilize the external data systems and take advantage of their powerful implementations (e.g. access methods and join algorithms)? 
+
+In this project, we experimented with a cross-database query system called Spark SQL, and we developed a cross-database query compiler that takes into account the 
+heterogeneity of the underlying infrastructure and it's able to optimize the query with respect to each table location.
+
+# What did you do?
+We mainly focused on the optimizer of our federated query compiler. We implemented a federated query compiler which is
+able to do optimize federated queries by splitting the execution of the initial query into individual, smaller parts
+that are pushed-down to the external database systems, achieving significant performance improvements. The optimizer
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md]: No hyphen in "pushed-down."

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+```sql
+SELECT AVG(A.salary)
+FROM A, B, C
+WHERE A.id = B.id
+AND A.id = C.id
+AND B.id = C.id
+```
+This query consists of a join between three tables: A, B and C. Let's assume three different data sources, namely DS1 
+and DS2 and assume the following table locations:
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:50]: Perhaps this is missing a listing of those locations?

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+In relational algebra, this join would look like the following: 
+`A ⋈ B ⋈ C`. By default, a federation engine (like Spark SQL) would first transfer all tables A, B and C to its system
+and then, it would execute the query its execution engine. However, more complex strategies could be followed for a query 
+like this. For example, `A ⋈ B ⋈ C` could be split into two complement subqueries, `A` (scan) and `B ⋈ C` (join), where 
+the first query which is just a scan would be executed as a simple scan from DS1. However, the second query (`B ⋈ C`)
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:60]: "is just a scan would be executed as a simple scan from DS1" seems duplicated.

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+a better performance compared to just fetching tables `B` and `C` to the federation engine (Spark SQL) and executing the 
+join there?
+
+## Learned Query Compilation
+To answer the previous question, we use a machine learning model as our cost model that helps us evaluate the cost of 
+a query without executing it. Using this model, we can easily evaluate whether it is better to push-down a query or not.
+In order to do that, we use a unified vector space model that is able to transform SQL queries into vectors, and thus,
+we can feed these queries to our model in order to train and predict the execution time of future queries. Thus, our
+optimizer is able to leverage this machine learning model in order to decide which parts of the query should push-down 
+or not.
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:71]: Two important questions this paragraph might raise:

* What is a "unified vector space model"? Is it a pre-trained embedding, like word2vec but for SQL queries?
* This is supervised learning, right? Like, the learned function is from SQL queries to latencies (in some specific unit, like seconds)? If so, when/how will it be trained (which is relevant to knowing when it will need to be retrained—is it every change to the data?)?

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+Let's take again a look at the previous example
+```sql
+SELECT AVG(A.salary)
+FROM A, B, C
+WHERE A.id = B.id
+AND A.id = C.id
+AND B.id = C.id
+```
+
+If we think of this query as a graph, it would look as follows:
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:99]: This could benefit from an explicit description of what the graph means. What is a vertex, and what is an edge?

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+```
+    A - B
+    \   / 
+      C
+```
+
+This is a fully-connected graph, as there are connections from all nodes to all nodes. As the final plan will be a binary tree,
+that means that the final plan could be any of the following three trees:
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:107]: Again, it would be nice to explicitly say that these trees represent the data flow between tables through operations (joins).

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+1.  ⋈         2.  ⋈          3.  ⋈ 
+   /  \          /  \           /  \
+  A   ⋈         B   ⋈          C   ⋈ 
+     / \           /  \           /  \
+    B   C         A    C         A    B
+```
+
+However, we are mostly interested in plans that look like 1. The reason is that we are mainly interested in joins that 
+can co-locate tables thus, so execution push-down is possible. Thus, we created a graph structure in which each node
+represents a table, and each edge represents a join between two tables. We enrich more our graph representation by adding
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:118]: Maybe this ("each node represents a table, and each edge represents a join between two tables") is the definition of the above graphs?

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+information about the location of each table in the node, that is, in which data source the node resides. This allows
+us to develop an algorithm can can parse this graph and generate possible trees that join tables that reside in the same
+engine, when that's possible given the input query semantics (joins).
+
+### Final Plan Generation
+After processing the graph, we get a tree that collocates tables residing in the same engine by producing a tree-plan
+that joins together the collocated tables. Then, we use our machine learning model in order to predict the execution cost
+of joining the collocated tables in the machine that they reside on in the federation engine. We then produce the final
+query execution plan which also contains information about where each node of the tree will be executed.
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:127]: I'm a little confused—what role does the learned cost model play then? This paragraph seems to say that it predicts the performance, but it doesn't say how that prediction is _used_. Are multiple plans generated and evaluated separately? How, in general, does the cost model guide the search for a good plan?

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+# What were the hardest parts to get right?
+The trickiest part of the project, which is still not completely resolved is the cost estimation. The average accuracy of 
+our model in predicting the cost of a given query is around 60%. However, in most of the cases in which we mispredict
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:131]: Define what you mean by "accuracy." L1 norm, normalized to some maximum cost, maybe?

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+# Evaluation
+In order to evaluate our system we use two benchmarks, [TPC-H](https://www.tpc.org/tpch/) and 
+[Join-Order-Benchmark](https://github.com/gregrahn/join-order-benchmark). The experimental setup consisted of
+one federation engine (Spark SQL) and two external systems, MySQL and PostgreSQL. 
+
+## Cost Model Accuracy
+We first try to evaluate the accuracy
+of our machine learning cost model. For this experiment, we evaluate how well the cost model does in selecting the
+appropriate execution engine (Spark SQL, Postgres or MySQL) in order to execute some part of the query. As mentioned in 
+the previous section, the cost model accuracy is around 60%. However, in this experiment our model mispredicts in cases
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:145]: I feel like it might be important to know the top-1 accuracy too: that is, not just how close the prediction is to the correct cost, but how often the predictor agrees with the ground-truth about the minimum-cost plan.

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+where the performance difference is very low. By running the TPC-H benchmark, we find that the average performance of
+running the queries in Spark without further optimization is 1150.57 ms. With further optimization and if we "manually"
+select the optimal execution engine each time, the optimal performance will be 122.666. Using the engine selections
+with our predictive model, the average query execution time is 152.7142, which is around the 1/7 of the query execution
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:149]: I would probably give these in seconds, not ms, and use much fewer significant figures (which currently imply more accuracy than you actually have). Like, 1.1 s, 0.12 s, and 0.15 s.

File: content/blog/2022-05-18-learned-federated-query-optimization.md [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+| Spark SQL | 1150.57 ms |
+| Optimal | 122.666 |
+| Predicted | 152.7142 |
+
+This experiment was the first simple proof-of-concept case in which all the tables reside in all engines,
+that is, both MySQL and SparkSQL. We run this experiment in order to demonstrate the accuracy of our cost model
+and to showcase that even in the case of multiple mispredictions, the final performance is close enough to the optimal.
+
+## Performance
+In order to measure the performance, we used the Join-Order-Benchmark. For those queries, we trained by changing the
```
Comment (sampsyo) [inline content/blog/2022-05-18-learned-federated-query-optimization.md:163]: Maybe it's just my databases naïveté, but I don' know what "the Join-Order-Benchmark" is.

**Review by sampsyo — CHANGES_REQUESTED** [review]
This seems really cool, @gsvic! Nice work getting this to work and outperform the SparkSQL baseline, even if there are more things to do here.

I have a few conceptual comments within. Please let me know if anything isn't clear.

File: content/blog/2022-05-18-learned-federated-query-optimization.md (hunk 1 above)
Comment (gsvic) [reply content/blog/2022-05-18-learned-federated-query-optimization.md]: sounds good, I will re-structure this.

File: content/blog/2022-05-18-learned-federated-query-optimization.md (hunk 4 above)
Comment (gsvic) [reply content/blog/2022-05-18-learned-federated-query-optimization.md]: done

File: content/blog/2022-05-18-learned-federated-query-optimization.md (hunk 5 above)
Comment (gsvic) [reply content/blog/2022-05-18-learned-federated-query-optimization.md]: done

File: content/blog/2022-05-18-learned-federated-query-optimization.md (hunk 2 above)
Comment (gsvic) [reply content/blog/2022-05-18-learned-federated-query-optimization.md]: done

File: content/blog/2022-05-18-learned-federated-query-optimization.md (hunk 3 above)
Comment (gsvic) [reply content/blog/2022-05-18-learned-federated-query-optimization.md]: done

File: content/blog/2022-05-18-learned-federated-query-optimization.md (hunk 5 above)
Comment (gsvic) [reply content/blog/2022-05-18-learned-federated-query-optimization.md]: done

File: content/blog/2022-05-18-learned-federated-query-optimization.md (hunk 1 above)
Comment (gsvic) [reply content/blog/2022-05-18-learned-federated-query-optimization.md]: done

File: content/blog/2022-05-18-learned-federated-query-optimization.md (hunk 9 above)
Comment (gsvic) [reply content/blog/2022-05-18-learned-federated-query-optimization.md]: done

**Review by gsvic — COMMENTED** [review]
First batch of updates. Fixes some types, restructured some sections. Fixed chapter names.

**Comment by sampsyo** [comment]
Hi! I know it's been a while, but would you be interested in wrapping this up so we can publish it?

**Comment by sampsyo** [comment]
Hello! I am closing old issues from last year's offering to prepare for the next iteration of the course. If you would like to have this blog post published, please just leave a note and we can reopen the thread.
